From: Laurence Evans <levans@solarflare.com>
Date: Thu, 25 Jan 2018 17:27:02 +0000
Subject: sfc: simplify RX datapath timestamping
Patch-mainline: v4.16-rc1
Git-commit: c4f64fcc4d31e7f773cb4eec9d90c40ebb049c14
References: bsc#1105555 FATE#326117

Use timestamp conversion function with correction to avoid duplicate
 correction handling.

Signed-off-by: Laurence Evans <levans@solarflare.com>
Signed-off-by: Edward Cree <ecree@solarflare.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 drivers/net/ethernet/sfc/ptp.c |   13 +++++--------
 1 file changed, 5 insertions(+), 8 deletions(-)

--- a/drivers/net/ethernet/sfc/ptp.c
+++ b/drivers/net/ethernet/sfc/ptp.c
@@ -586,8 +586,6 @@ static int efx_ptp_get_attributes(struct
 		return -ERANGE;
 	}
 
-	ptp->time_format = fmt;
-
 	/* MC_CMD_PTP_OP_GET_ATTRIBUTES is an extended version of an older
 	 * operation MC_CMD_PTP_OP_GET_TIME_FORMAT that also returns a value
 	 * to use for the minimum acceptable corrected synchronization window.
@@ -1869,10 +1867,7 @@ void __efx_rx_skb_attach_timestamp(struc
 	u32 diff, carry;
 	struct skb_shared_hwtstamps *timestamps;
 
-	pkt_timestamp_minor = (efx_rx_buf_timestamp_minor(efx,
-							  skb_mac_header(skb)) +
-			       (u32) efx->ptp_data->ts_corrections.rx) &
-			      (MINOR_TICKS_PER_SECOND - 1);
+	pkt_timestamp_minor = efx_rx_buf_timestamp_minor(efx, skb_mac_header(skb));
 
 	/* get the difference between the packet and sync timestamps,
 	 * modulo one second
@@ -1910,8 +1905,10 @@ void __efx_rx_skb_attach_timestamp(struc
 
 	/* attach the timestamps to the skb */
 	timestamps = skb_hwtstamps(skb);
-	timestamps->hwtstamp =
-		efx_ptp_s27_to_ktime(pkt_timestamp_major, pkt_timestamp_minor);
+	timestamps->hwtstamp = efx_ptp_s27_to_ktime_correction(
+				pkt_timestamp_major,
+				pkt_timestamp_minor,
+				efx->ptp_data->ts_corrections.rx);
 }
 
 static int efx_phc_adjfreq(struct ptp_clock_info *ptp, s32 delta)
